### ISSUE TITLE
RNNEncoder使用位置问题

### ISSUE THREAD

**Issue by SuFame920** [TheShy-Dream/AtCAF#10]
作者您好，再次感激您对MSA做出的贡献。在复现您的成果中，我也学习到了很多。
另外，我对encoders.py中RNNEncoder的使用位置存在一个疑问。在论文中作者你是先将音视频特征经过LSTMs处理后再用Conv1D进行特征重映射，我仔细观察您的代码后发现一个细节上的问题即在去偏之前貌似并没有使用RNNEncoder，而是在自注意力、跨注意力模块中各自仅执行了Conv1D映射。同时，我也观察到RNNEncoder是在反事实模块后对原本的未去偏的visual和acoustic编码的，编码后是去计算互信息了。以上是我的疑问，若作者您方便的时候希望您能为我解疑！再次感谢！

Code：
if self.training:
            # fusion=fusion-counterfactual_fusion
            text = text_trans[0,:,:]  # 32*768 (batch_size, emb_size)
            acoustic = self.uni_acoustic_enc(acoustic, a_len)  # 32*16   #<-------- 第一次使用RNNEncoder
            visual = self.uni_visual_enc(visual, v_len)  # 32*16               #<-------- RNNEncoder

            if y is not None:
                lld_tv, tv_pn, H_tv = self.mi_tv(x=text, y=visual, labels=y, mem=mem['tv'])
                lld_ta, ta_pn, H_ta = self.mi_ta(x=text, y=acoustic, labels=y, mem=mem['ta'])
                # for ablation use
                if self.add_va:
                    lld_va, va_pn, H_va = self.mi_va(x=visual, y=acoustic, labels=y, mem=mem['va'])
            else:  # 默认进这
                lld_tv, tv_pn, H_tv = self.mi_tv(x=text, y=visual)  # mi_tv 模态互信息
                # lld_tv:-2.1866  tv_pn:{'pos': None, 'neg': None}  H_tv:0.0
                lld_ta, ta_pn, H_ta = self.mi_ta(x=text, y=acoustic)
                if self.add_va:
                    lld_va, va_pn, H_va = self.mi_va(x=visual, y=acoustic)

            nce_t = self.cpc_zt(text, fusion)  # 3.4660
            nce_v = self.cpc_zv(visual, fusion)  # 3.4625
            nce_a = self.cpc_za(acoustic, fusion)  # 3.4933

            nce = nce_t + nce_v + nce_a  # 10.4218  CPC loss

            pn_dic = {'tv': tv_pn, 'ta': ta_pn, 'va': va_pn if self.add_va else None}
            lld = lld_tv + lld_ta + (lld_va if self.add_va else 0.0)  # -5.8927
            H = H_tv + H_ta + (H_va if self.add_va else 0.0)

<img width="345" height="74" alt="Image" src="https://github.com/user-attachments/assets/a96915a1-fe49-431e-933d-da6db097e500" />

**Comment by SuFame920** [comment]
如果可能，mosi_data_noalign.pkl 和 mosei_senti_data_noalig.pkl中的音视频特征是已经经LSTMs处理后的时序特征？

**Comment by TheShy-Dream** [comment]
您好，感谢您对我们的关注。为了方便，我统一回复您的上述两个问题了哈。
Q1：RNN的处理位置
A1：您的观察是正确的，与原文存在些许出入，RNN用在自注意力以后，经过RNN的特征不进行Conv1D以及模态融合，直接进行互信息计算。因为大家其实都是这么操作，然后画的时候也是画成我那样。可以参考TeFNA，TeTFN，原因很简单。因为LSTM处理以后一般取最后一个时间步，因此维度直接变成[bs,dim]dim为LSTM的hidden_dim，这时候再用Conv1D再融合就不对了，因为注意力的计算要考虑一个序列的，现在序列维度已经不见了。另外，我在自注意力和交叉注意力中都使用了Conv1D，这个只是类似MLP的作用，无需在意。
Q2：mosi_data_noalign.pkl 和 mosei_senti_data_noalig.pkl中的音视频特征是已经经LSTMs处理后的时序特征？
A2：mosi_data_noalign.pkl 和 mosei_senti_data_noalig.pkl这两个文件存储的是特征预处理(librosa,openface等提取器直接提取)以后一些信息，并不是LSTMs处理后的时序特征。

如果有其他疑问，欢迎继续放在issue区，我们继续探讨。再一次感谢您对我们工作的关注，祝您科研顺利。

**Comment by SuFame920** [comment]
谢谢！
